src/core/src/op/convert_align_types.cpp

```diff
@@ -0,0 +1,171 @@
+                                  element::u4,
+                                  element::u8,
+                                  element::u16,
+                                  element::u32,
+                                  element::u64,
+                                  element::f8e4m3,
+                                  element::f8e5m2,
+                                  element::bf16};
+    const auto& lhs_type = lhs.get_element_type();
+    const auto& rhs_type = rhs.get_element_type();
+    NODE_VALIDATION_CHECK(op,
+                          std::find(supported_types.begin(), supported_types.end(), lhs_type) != supported_types.end());
+    NODE_VALIDATION_CHECK(op,
+                          std::find(supported_types.begin(), supported_types.end(), rhs_type) != supported_types.end());
+    if (lhs_type.is_dynamic() || rhs_type.is_dynamic()) {
+        return element::dynamic;
+    }
+    if (lhs_type == rhs_type)
+        return lhs_type;
+    if (lhs_type == element::boolean)
+        return rhs_type;
+    if (rhs_type == element::boolean)
+        return lhs_type;
+    const auto& lhs_rank = lhs.get_partial_shape().rank();
+    const auto& rhs_rank = rhs.get_partial_shape().rank();
+    const bool is_lhs_scalar = lhs_rank.is_static() ? lhs_rank.get_length() == 0 : false;
+    const bool is_rhs_scalar = rhs_rank.is_static() ? rhs_rank.get_length() == 0 : false;
+    const bool is_lhs_signed = lhs_type.is_signed();
+    const bool is_rhs_signed = rhs_type.is_signed();
+    const bool is_lhs_real = lhs_type.is_real();
+    const bool is_rhs_real = rhs_type.is_real();
+    const size_t lhs_bitwidth = lhs_type.bitwidth();
+    const size_t rhs_bitwidth = rhs_type.bitwidth();
+
+    if (pytorch_scalar_align && (is_lhs_scalar ^ is_rhs_scalar) && (is_lhs_real == is_rhs_real)) {
+        if (promote_unsafe) {
+            return is_lhs_scalar ? rhs_type : lhs_type;
+        }
+        const auto target = is_lhs_scalar ? rhs_type : lhs_type;
+        const auto scalar = is_lhs_scalar ? lhs_type : rhs_type;
+        if ((target.is_signed() == scalar.is_signed() && target.bitwidth() >= scalar.bitwidth()) ||
+            (target.is_signed() && !scalar.is_signed() && target.bitwidth() * 2 >= scalar.bitwidth())) {
+            return target;
+        }
+        NODE_VALIDATION_CHECK(op, false, " Scalar input cannot be PyTorch-like aligned using safe promotion rules.");
+    }
+    if (is_lhs_real ^ is_rhs_real) {
+        if (promote_unsafe) {
+            return is_lhs_real ? lhs_type : rhs_type;
+        }
+        const auto real = is_lhs_real ? lhs_type : rhs_type;
+        const auto integer = is_lhs_real ? rhs_type : lhs_type;
+        if (real.bitwidth() >= integer.bitwidth() * 2) {
+            return real;
+        }
+        NODE_VALIDATION_CHECK(op, false, "Integer input cannot be safely promoted to floating-point.");
+    }
+    if ((is_lhs_real == is_rhs_real) && (is_lhs_signed ^ is_rhs_signed)) {
```

Comment:
Not sure what this check mean e.g. if both are floating points the condition is always false as floating points are always signed.
Add some explanation or use different flags. Check other cases.

Reply:
Those could be either both floating or both integer, added some comments to code to better explain it